### DISCUSSION TITLE
[English] Help Required : Input Method For a Non CJK, Non-Logographic Script (The Sinhalese Script)

### DISCUSSION THREAD

**Discussion by AdelinaM17n** [rime/home#1729] [Q&A] [answered]
Hello dear Rimers, I've been eyeing RIME for a while as an avenue to implement my transliteration based [input method for Sinhala](https://github.com/AdelinaM17n/Sinhala-Addendum-Phonetic-M17n/blob/main/si-addendum.mim) on top of the Rime Engine but I've hit one major roadblock which is that I am not a speaker of Chinese languages, as such I've not been able to properly read the rime documentation in the wiki section. Even with the help of Google translate it was still hard to read and understand the documentation.

## Why Would I Use Rime For This ?
You might have noticed that the input method repository I've linked already contains an implementation of it with the m17n library, something much better suited for a transliteration input method like this and so you might be wondering as to why I am bothering to implement this with Rime at all.
   
The major reason is that RIME offers is that its stellar support for a wide ranging selection of platforms, if I am able to get this input method implemented with RIME, I'd be able to effortlessly bring my input method to many platforms.
   
There, however, is another cross platform input method engine that is much more suited transliteration-based IMEs, that being Keyman and in fact I've already [implemented this IME with the keyman engine](https://github.com/AdelinaM17n/Sinhala-Addendum-Phonetic-Keyman), but Keyman's windows implementation still appears to have few compatibility hiccups with apps such as Discord, which makes it sub-optimal until those issues are resolved.
   
So this leaves me with RIME as the best and only option besides individually implementing the IME for every platform.

## Things I Already Tried With RIME / The Issues I Ran Into
In my search for reference implementations of IMEs with RIME, I ran into this [Tamil input method](https://github.com/AraragiHoozuki/Tamil-for-Rime), as the Tamil script (as with all the other Indic scripts) is structurally very similar to the Sinhalese script, this was a very promising start. I was able to put together a bare-bones input method using that as a starting point, but there were still many pain points that left me desiring for a more robust implementation, if possible.
   
The main issue I faced with that was that clicking the `space` key would commit the text in the preview without inputting a space. While this makes sense for logographic writing systems (such as those of CJK languages), it is frustrating and inconvenient for an alpha-syllabic writing system like Sinhala.
- ***So if anyone knows how to circumvent this issue, that alone would be enough to have a passable implementation of this input method***

There are many things I'd ideally do differently than this specific implementation/reference point, such as
- Display the current output text to be committed within the pre-edit text, if possible.
- Commit the text in the pre-edit programmatically from within the IME without having the user tap the space key (Hard to explain in one line, and it makes much more sense with an alpha-syllabic writing system than with a logographic one).   

But those are niceties that I could live without, if it's not possible to achieve that with RIME 

## I can't come up with a title for this section
So any help would be appreciated, whether that be links to English language documentation or reference implementations I could use or any bit of advice or help you could offer.
   
If you want to directly contact me 
- Discord - m_adelina
- Email(s) - `MaheeshaAdelina@Gmail.com`, `AdelinaFM17n@Gmail.com`

**Comment by lotem** [comment, marked as answer]
Hello.

I have answers to some of your questions since people have tried to use Rime for other languages/scripts.

## Space key

The option `punctuator {use_space: true}` might be what you need.

Use this code snippet.
https://github.com/biopolyhedron/rime-latin-international/blob/9166ff00571cf9fd2088f9613a9f91058c5b2a43/latin_international.schema.yaml#L59

Do include `half_shape: { ' ' : { commit: ' ' } }` but don't write `&punct` and `*punct` since you don't need the `full_shape` section anyway.

## Display the current output text as pre-edit

The `librime` library exposes both information to the IME front-end application, and it's up the app to decide what to present.
For example the Windows client - Weasel has this configuration option (`weasel.yaml` is for front-end behaviours):
[`style/preedit_type`](https://github.com/rime/weasel/blob/37c8fa161a221a263bb439a1158ba401c0cf90a3/output/data/weasel.yaml#L35)

I found this in Weasel's changelog (translated):

    New option `preedit_type: preview_all`, display the candidate text as inline composition.
    style/preedit_type: "composition" | "preview" | "preview_all"

I think you need `preview_all`. The difference between `preview` and `preview` all might be, often in typing Chinese, you can partially convert a sentence by selecting a word in the middle, in that case `preview` displays the next word inline, and `preview_all` displays the partially converted sentence up to next word.

## Commit the text in the pre-edit programmatically

Use the option [`speller: { auto_select: true }`](https://github.com/biopolyhedron/rime-latin-international/blob/9166ff00571cf9fd2088f9613a9f91058c5b2a43/latin_international.schema.yaml#L49)

This commits the current candidate when there is no ambiguity - either it's the unique candidate at the moment, or the first candidate when the next key press does not form a longer valid code.

The key to make this option work is that you must also disable [`translator: { enable_sentence: false }`](https://github.com/biopolyhedron/rime-latin-international/blob/9166ff00571cf9fd2088f9613a9f91058c5b2a43/latin_international.schema.yaml#L54) otherwise longer code are always made valid (with multi-character candidates).

Make sure you have `translator: { enable_completion: true }` (which is the default); this shows candidates for next-key prediction (for example when typing `e`, the candidate list also includes the character for `ee`). This avoids committing "unique candidate" too early when it can be part of a longer input code.

**Reply by AdelinaM17n** [reply]
Thanks for the quick and detailed response, this is really helpful

**Reply by AdelinaM17n** [reply]
Also one thing to note, it seems that `style/preedit` value that best suits my needs seems to `preview`, `preview_all` seems to display all the possible candidates in-line within the preedit.

One other minor thing, is it possible for it have it so that the candidate selection menu wouldn't show up?

**Comment by lotem** [comment]
## About documenting in Chinese

The majority user of Rime speak and read Chinese, which has been the focus of the project. The library implement algorithms and data structures that are - while not strictly coupled with - optimized for the Chinese language/script.

The situation is that Chinese users don't get adequate documentation since the code including configuration code are written in English. I owe you a sorry but all other languages combined only makes up a small percentage of Rime users and the need for documentation. As Chinese input method will still be the focus of the project, I'm leaning to write more code, configuration and documentation in Chinese in future development.

**Reply by AdelinaM17n** [reply]
Yea it's understandable that you are mainly targeting speakers of the Chinese language. And the community for IMEs for non CJK languages is very very niche that it's understandable why you don't want to focus on that anytime soon.
